src/dinit.cc

```diff
@@ -453,6 +471,87 @@ static int process_commandline_arg(char **argv, int argc, int &i, options &opts)
     return 0;
 }
 
+#if SUPPORT_SELINUX
+// Load the system SELinux policy and transition ourselves to it. When successful, this will cause
+// SELinux labels as per the policy to be attached to processes (and file descriptors owned by
+// those processes). The SELinux framework will begin to enforce restrictions on access based on
+// these labels and the loaded policy.
+//
+// We might lose access to any file descriptors we have open when this is called (since they will
+// still be labelled with the kernel context), so it is best done early (i.e. before we start
+// opening file descriptors).
+//
+// Parameters:
+//   exe - the path that we are invoked with (to calculate our new security context to transition into.)
+//
+// Returns:
+//   If we fail to load the system SELinux policy, return false, otherwise, return true.
+static bool selinux_transition(const char *exe) {
+    // Let's use std::cerr instead of the log for logging messages here.
+    // If we output anything, we return failure, which indicates dinit should
+    // terminate, which may happen before the log is initalised and flushed.
+    using std::cerr;
+    using std::endl;
+
+    char *current_context = nullptr;
+    char *file_context = nullptr;
+    security_class_t security_class;
+    char *new_context = nullptr;
+
+    if (is_selinux_enabled() == 1) {
+        return true;
+    }
+
+    int enforce = 0;
+    // We don't need to worry about the enforcing=0 kernel cmdline option or parsing
+    // /etc/selinux/config, selinux_init_load_policy(3) will handle all cases for us.
+    if (selinux_init_load_policy(&enforce) != 0) {
+        if (enforce > 0) {
+            cerr << "Failed to load SELinux policy." << endl;
+            return false;
+        }
```

Comment:
So if the load fails (`selinux_init_load_policy` returns 0) but `enforce` is 0, what is supposed to happen here? It seems like it just ploughs ahead with trying to set the context, but surely that isn't right?

Reply:
For the `enforcing` variable, there are 2 states:
- = 0 is permissive mode
- > 0 is enforcing mode

In the case of enforcing, the system is expected to boot in a secure state, i.e. with the system's SELinux policy enforcing. If we fail to load it, we should halt the boot as that is an untrusted state.

In the case of permissive, it's generally better to be more lenient. In permissive mode, actions are expected to be logged, but any failures revolving around the policy aren't expected to have an impact on the system's operation. Other SELinux aware applications, such as `load_policy(1)` from policycoreutils also reflect this behavior:
https://github.com/SELinuxProject/selinux/blob/main/policycoreutils/load_policy/load_policy.c#L68-L77

Let me know your thoughts, I can add a comment explaining this too if desired

Reply:
That's not answering the question I asked. If the policy fails to load (`selinux_init_load_policy` returns 0) but `enforce` is 0, what is this function supposed to do?

The comment says:
> If we fail to load the system SELinux policy, return false

But it's not even returning at this point, it's just going on through to the following call to `(getcon_raw(&current_context)`. Isn't that wrong?

> Other SELinux aware applications, such as load_policy(1) from policycoreutils also reflect this behavior

That code checks whether loading the policy failed, and if so, prints an error message and retuns an exit code:

https://github.com/SELinuxProject/selinux/blob/main/policycoreutils/load_policy/load_policy.c#L82-L85

That's different to what is here, which just continues as if no error had happened.

Reply:
And when I say "continues as if no error had happened", I mean that it goes on to try and set a new context (`getcon_raw`, `getfilecon_raw`, `security_compute_create_raw`, etc). I assume that those will fail, since the policy load failed, right?

Reply:
```
I assume that those will fail, since the policy load failed, right?
```
Those won't fail as it'll just continue on with the kernel context.

Reply:
>But it's not even returning at this point, it's just going on through to the following call to  (getcon_raw(&current_context). Isn't that wrong?
I see what you meant now. Nothing should be inherently wrong with that, but I do agree it just makes sense to return (probably true) in this case. We can print out an error, return true early (as not to exit) and continue. Would that work?

> That's different to what is here, which just continues as if no error had happened.
Ah I see, thought I'd logged that.

Reply:
I've pushed https://github.com/davmac314/dinit/pull/400/commits/66be73d68337800a1d26c8a3fb186957fa92c835 that should fix this. Let me know your thoughts, thanks :)

Reply:
> Those won't fail as it'll just continue on with the kernel context

Are you saying that `setcon_raw` will do nothing but still return success?

Reply:
> Let me know your thoughts

I know that you have done some work on this and are probably trying hard, but in all honesty I am very close to pulling the plug. I need you to clearly answer the questions that I'm asking, not just push more code.

The question I asked was:

> So if the load fails (selinux_init_load_policy returns 0) but enforce is 0, what is supposed to happen here? It seems like it just ploughs ahead with trying to set the context, but surely that isn't right?

I still don't have a satisfactory answer to that. I can't accept this PR without understanding what it is doing.

Reply:
> > Those won't fail as it'll just continue on with the kernel context
> 
> Are you saying that `setcon_raw` will do nothing but still return success?

I decided to clarify this a bit further, so `getcon_raw` seems to be fine (which makes logical sense), and `setcon_raw` will be fine, but the issue lies in `security_compute_create_raw` as that actually attempts to read the *policy* to calculate something, so that should fail. I tested it by disabling the load temporarily on my machine, and writing a small program to see what would/wouldn't fail regarding those functions. So logically based off all of this, I think the early return makes sense.

Reply:
> > Let me know your thoughts
> 
> I know that you have done some work on this and are probably trying hard, but in all honesty I am very close to pulling the plug. I need you to clearly answer the questions that I'm asking, not just push more code.
> 
> The question I asked was:
> 
> > So if the load fails (selinux_init_load_policy returns 0) but enforce is 0, what is supposed to happen here? It seems like it just ploughs ahead with trying to set the context, but surely that isn't right?
> 
> I still don't have a satisfactory answer to that. I can't accept this PR without understanding what it is doing.

Understood. So:
```
So if the load fails (selinux_init_load_policy returns 0) but enforce is 0, what is supposed to happen here? It seems like it just ploughs ahead with trying to set the context, but surely that isn't right?
```
This is accurate, i.e. that wasn't right. I don't think we should fail (as in exit) there, and I incorrectly gathered that it would just fall through (because it'll attempt to set the same context). I overlooked the fact `security_compute_create_raw` specifically reads from the policy store, so that'd fail. So an early return of true would make sense. Sincere apologies

Reply:
> Sincere apologies

Ok, but: you need to _slow down_ and read my questions/comments properly, and answer them comprehensively. I'm spending far more time on this that I would like, and it seems like it's mostly because I have to repeatedly ask questions until I get a straight answer.

> So an early return of true would make sense

"An early return of true" will currently fail the boot. So is that what is supposed to happen?

> so getcon_raw seems to be fine (which makes logical sense),

Where does this information come from? (I'm having trouble even finding the source for getcon_raw).

Reply:
> Ok, but: you need to slow down and read my questions/comments properly, and answer them comprehensively. I'm spending far more time on this that I would like, and it seems like it's mostly because I have to repeatedly ask questions until I get a straight answer.

Noted, I'll make sure to, apologies once again.

> "An early return of true" will currently fail the boot. So is that what is supposed to happen?
I don't think that's correct, unless I'm missing something:
```c++
    if (am_system_mgr && am_system_init && opts.load_selinux_policy && !selinux_transition(argv[0])) return 1;
```
Would `!selinux_transition(argv[0])` mean that the function would need to return false for it to return 1 and hence fail the boot here?

> Where does this information come from? (I'm having trouble even finding the source for getcon_raw).
This came from a local test (i.e. by not loading the policy on my system and seeing if getcon_raw fails). This makes sense as well, because from the manpage:
```
getcon_raw(), getprevcon_raw(), getpidcon_raw(), getpeercon_raw() and setcon_raw() behave identically to their non-raw counterparts but do not perform context translation.
```
No context translation would mean that the policy store is not needed. The raw context of any PID is also available in /proc/1/attr/current, and without loading the policy I can also cat that. Those return kernel without the policy being loaded for me.

Reply:
Sorry: an early return of true won't fail the boot, I was thinking an early return of false.
However, an early return of true would break the contract described in the function comment:
```
// Returns:
//   If we fail to load the system SELinux policy, return false, otherwise, return true.
```
So that might need to be updated, or expanded on in the descriptive part of the comment.